python/tvm/relay/quantize/quantize.py

```diff
@@ -0,0 +1,161 @@
+import math
+import numpy as np
+
+from . import _quantize
+from .. import expr as _expr
+from .. import ir_pass as _ir_pass
+from .. import op as _op
+from .. import build_module as _build
+from ..base import register_relay_node
+from ..._ffi.function import register_func
+
+
+class QFieldKind(object):
+    INPUT = 1
+    WEIGHT = 2
+    ACTIVATION = 3
+
+
+class QConfig(object):
+    current = None
+
+    def __init__(self, bit_dict=None, global_scale=8.0, skip_k_conv=0):
+        if bit_dict is None:
+            bit_dict = {
+                QFieldKind.INPUT: 8,
+                QFieldKind.WEIGHT: 8,
+                QFieldKind.ACTIVATION: 32
+            }
+        self.bit_dict = bit_dict
+        self.global_scale = global_scale
+        self.skip_k_conv = skip_k_conv
+        self.counter = 0
+
+    def __enter__(self):
+        self._old_scope = QConfig.current
+        QConfig.current = self
+        return self
+
+    def __exit__(self, ptype, value, trace):
+        assert self._old_scope
+        QConfig.current = self._old_scope
+
+
+QConfig.current = QConfig()
+
+
+def qconfig(bit_dict=None, global_scale=2.0, skip_k_conv=0):
+    return QConfig(bit_dict, global_scale, skip_k_conv)
+
+
+def get_current_qconfig():
+    return QConfig.current
+
+
+@register_func("relay.quantize.get_config_bit")
+def get_config_bit(kind):
+    cfg = get_current_qconfig()
+    return cfg.bit_dict[kind]
+
+
+@register_relay_node
+class QFieldExpr(_expr.TempExpr):
+    def __init__(self, expr, kind):
+        self.__init_handle_by_constructor__(
+            _quantize.make_qfield_expr, expr, kind)
+
+
+def simulated_quantize(data, dom_scale, bit, clip_min, clip_max,
+                       sign, rounding, kind):
+    return _quantize.simulated_quantize(data, dom_scale, bit, clip_min, clip_max,
+                                        sign, rounding, kind)
+
+
+@register_func("relay.quantize.attach_simulated_quantize")
+def attach_simulated_quantize(data, kind):
+    dom_scale = _expr.var("dom_scale")
+    bit = _expr.var("bit")
+    clip_min = _expr.var("clip_min")
+    clip_max = _expr.var("clip_max")
+    return simulated_quantize(data, dom_scale, bit, clip_min, clip_max,
+                              True, "round", kind)
+
+
+def register_qfield_rewrite(op_name, frewrite=None, level=10):
+    return _op.register(op_name, "FQFieldRewrite", frewrite, level)
+
+
+def annotate(graph):
+    """
+    `annotate` will construct a simulated quantized graph
+    center around some ops like `conv` and broadcast the quantize part
+    """
+    return _quantize.annotate(graph)
+
+
+def calibrate(graph, dataset=None):
+    """
+    `calibrate` will find a set of suitable bit and scale,
+    also make them as constant in graph
+    """
+    def _scalar(x, dtype):
+        return _expr.const(np.array(x).astype(dtype))
+
+    # weight scale with nearest mode-2 scale
+    def power2_scale(arr):
+        val = np.amax(np.abs(arr.asnumpy()))
+        assert val > 0
```

Comment:
Maybe it makes sense to set the scale to 1.0 if val == 0, otherwise it throws an error when we have default initialized params with bias=0